src/lammpsplugin.cpp

```diff
@@ -0,0 +1,117 @@
+/* ----------------------------------------------------------------------
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   https://lammps.sandia.gov/, Sandia National Laboratories
+   Steve Plimpton, sjplimp@sandia.gov
+
+   Copyright (2003) Sandia Corporation.  Under the terms of Contract
+   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+   certain rights in this software.  This software is distributed under
+   the GNU General Public License.
+
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+#include "lammpsplugin.h"
+
+#include "error.h"
+#include "force.h"
+#include "lammps.h"
+#include "modify.h"
+
+#include <vector>
+
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <dlfcn.h>
+#endif
+
+namespace LAMMPS_NS
+{
+  static std::vector<lammpsplugin_entry_t> pluginlist;
```

Comment:
@akohlmey shouldn't this be an instance variable of LAMMPS or of a child object? E.g. `LAMMPS::Plugins`? These methods also look like they should become part of the library interface and follow the same naming conventions: `lammps_plugin_load`, `lammps_plugin_register`.

Reply:
nope. I have thought about that.  Whether those are loaded or not is a global property of the executable/shared library. I was actually thinking the other way around, that the maps for the individual styles should also be global lists and not part of the objects that create them. Adding the plugin system is making things a bit more complicated and adds new problems like "should we allow to "unload" statically loaded classes?" or "how do we avoid leaking dso handles?", but it can also be useful for development since you can load new functionality without having to rebuild the full executable/library and you can replace existing styles with new implementations without rebuilding.

Reply:
> Then it needs some additional protections.

Certainly. This is a bit of a "you know it when you see it" situation, where you first need to figure out a way to do something, then study its utility and usability and then refine or redo.  Which is why I make this public now while I am still exploring. This must be the 3rd time I am starting to looking into this seriously, and now is the first time I am getting somewhere that looks like it can work the way I expect it to work. One thing I am trying to avoid now is to over-engineer the initial version. Previously, I would always start with something too complex; now I am trying to wait.

If you want, we can discuss some more about this and how to manage the "factory maps" person-to-person.